feat: teams settings editor [BD-38] [TNL-8453] [BB-4843]

xitij2000 opened the pull request:
Adds a new editor for the team sets, completing all the apps for pages and resources.

**JIRA tickets**:  TNL-8101 [TNL-8453](https://openedx.atlassian.net/browse/TNL-8453)

**Screenshots**: 


https://user-images.githubusercontent.com/118837/132700674-ee7c8d45-8402-459d-aa2e-c3ce15139f0a.mp4



**Testing instructions**:

1. For a course that doesn't have teams configured, visit http://localhost:2001/course/course-v1:edX+DemoX+Demo_Course/pages-and-resources/teams/settings
2. Add a new team set
3. Check that the validation works properly
4. Check that you can add teams
5. Check that you can delete teams
6. Check that you can edit teams
7. Check that the teams show up in the LMS at http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course/teams/


**Reviewers**
- [ ] (OpenCraft internal reviewer's GitHub username goes here)
- [ ] edX reviewer[s] TBD

edx-webhook commented:
Thanks for the pull request, @xitij2000! I've created [BLENDED-956](https://openedx.atlassian.net/browse/BLENDED-956) to keep track of it in Jira.   More details are on the [BD-38](https://openedx.atlassian.net/wiki/spaces/COMM/pages/2790490201/BD-38+Discussions+In-Context+Discussion+Moderation+Phase+2) project page.


When this pull request is ready, tag your edX technical lead.


<!-- comment:welcome-blended -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

xitij2000 commented on src/pages-and-resources/teams/Settings.jsx:
```diff
@@ -0,0 +1,105 @@
+  const handleSettingsSave = (values) => {
+    // For newly-added teams, fill in an id.
+    const topics = values.topics.map(topic => ({
+      id: topic.id || uuid(),
```
Instead of a UUID perhaps it might be better to slugify the topic name? This will result in friendlier URLs.

meysam81 replied:
@xitij2000 

I don't think it makes much of a difference overall. But perhaps it wouldn't be a bad idea either. I'd rather the slug most probably.

xitij2000 replied:
The difference it makes is that this ID is used in URLs. So in one case users see a URL like '/north-america' and in the other '/6432aabb-0829-4ca5-b9b5-83829c94f277'.

xitij2000 replied:
Actually this might be harder said than done, especially if you consider unicode. I think the best approach is a uuid or a manual slug.

meysam81 commented on src/pages-and-resources/teams/TeamSetEditor.jsx:
```diff
@@ -0,0 +1,193 @@
+ * @type {{MIN: number, MAX: number, DEFAULT: number}}
+ */
+export const TeamSizes = {
+  DEFAULT: 50,
+  MIN: 5,
+  MAX: 500,
```
Shouldn't these be configurable from the deployment section as well? Something like reading from the environment. As explained in [12factorapp](https://12factor.net/config).

xitij2000 replied:
So currently these are not configurable in edx-platform either, so I don't see the point of making them configurable here.  I've separated them out so that they can be made configurable in the future. However, simply making them configurable here without a corresponding backend change won't help much.

mehaknasir commented on src/pages-and-resources/teams/TeamSetEditor.jsx:
```diff
@@ -0,0 +1,193 @@
+                  onBlur={onBlur}
+                />
+                <Form.Text>{intl.formatMessage(messages.teamSetFormNameHelp)}</Form.Text>
+                {(touched?.name && errors?.name) && (
+                <Form.Control.Feedback type="invalid" hasIcon={false}>
+                  {intl.formatMessage(messages.teamSetFormNameError)}
+                </Form.Control.Feedback>
+                )}
```
for all the error messages I would simplify it by creating a generic component for example

```
function ErrorMessage({ errors, touched,  name }) {
  if (!errors[name] || !touched[name]) {
    return null;
  }
  return <p>{errors[name]}</p>;
}
```
and in this case, it can be modified as 

```
function ErrorMessage({ errors, touched,  name, key }) {
  if (!errors[name] || !touched[name]) {
    return null;
  }
  return <p>{intl.formatMessage(messages[key])}</p>;
}
```
and usage will be like 

`<ErrorMessage errors={errors} touched={touched}  name={name}  key='teamSetFormNameError' />`

but if we transfer error messages to yup validations it will get shorten like in the first example. and I will suggest shifting error messages with yup validations.

awais-ansari reviewed (changes requested):
UI is not matching with Figma. The errors transition effect is not working also.  You can have a look at the [discussion topic](https://github.com/edx/frontend-app-course-authoring/blob/master/src/pages-and-resources/discussions/app-config-form/apps/shared/discussion-topics/DiscussionTopics.jsx). 
<img width="477" alt="Screenshot 2021-09-08 at 10 57 48 PM" src="https://user-images.githubusercontent.com/79941147/132560431-919fea1f-0973-493c-b31d-d62a6c3ba348.png">

<img width="284" alt="Screenshot 2021-09-08 at 10 56 54 PM" src="https://user-images.githubusercontent.com/79941147/132560293-bda39643-40b4-4419-a079-6e897f3cc40c.png">

xitij2000 commented:
@awais-ansari  I was working from a different older version of the mockup, hence the differences. I have updated the code now to extract common code between this UI and the discussion topic edit.

xitij2000 commented on src/pages-and-resources/teams/TeamSetEditor.jsx:
```diff
@@ -0,0 +1,193 @@
+          : (
+            <React.Fragment key="isConfiguring">
+              <Form.Group key="isConfiguring">
+                <Form.Control
+                  name={`${fieldNameCommonBase}.name`}
+                  floatingLabel={intl.formatMessage(messages.teamSetFormNameLabel)}
+                  defaultValue={teamSet.name}
+                  onChange={onChange}
+                  onBlur={onBlur}
+                />
```
I've added a uniqueness check here.

meysam81 reviewed (commented):
How can I test this in local dev?

xitij2000 commented on package.json:
```diff
@@ -36,46 +36,46 @@
-    "@reduxjs/toolkit": "1.5.0",
-    "classnames": "2.2.6",
-    "core-js": "3.8.1",
+    "@edx/frontend-component-footer": "10.1.6",
```
They have to go with some PR. I needed to upgrade some dependencies for this so just did a full upgrade.

awais-ansari commented on src/pages-and-resources/teams/Settings.jsx:
```diff
@@ -0,0 +1,150 @@
+      initialValues={{ ...teamsConfiguration }}
+      validationSchema={{
+        maxTeamSize: Yup.number()
+          .required(intl.formatMessage(messages.maxTeamSizeEmpty))
+          .min(TeamSizes.MIN)
+          .max(
+            TeamSizes.MAX,
+            intl.formatMessage(messages.maxTeamSizeTooHigh, {
+              max: TeamSizes.MAX,
+            }),
+          ),
+        topics: Yup.array().of(
+          Yup.object({
+            id: Yup.string(),
+            name: Yup.string()
+              .required(intl.formatMessage(messages.teamSetFormNameEmpty))
+              .trim(),
+            type: Yup.string().oneOf(Object.values(TeamSetTypes)),
+            description: Yup.string()
+              .required(intl.formatMessage(messages.teamSetFormDescriptionError))
+              .trim(),
+            maxTeamSize: Yup.number()
+              .min(TeamSizes.MIN)
+              .max(
+                TeamSizes.MAX,
+                intl.formatMessage(messages.maxTeamSizeTooHigh, {
+                  max: TeamSizes.MAX,
+                }),
+              )
+              .default(null),
```
Global max team size validation is not working when a user tries to submit a team form without touch the max team size.
https://loom.com/share/7769189213cc4e10bd2d2338ee8e281b

awais-ansari commented on src/pages-and-resources/teams/Settings.jsx:
```diff
@@ -0,0 +1,150 @@
+        }) => (
+          <>
+            <div>
+              <h4 className="mb-3">{intl.formatMessage(messages.teamSize)}</h4>
+              <Form.Group>
+                <Form.Control
+                  type="number"
+                  name="maxTeamSize"
+                  floatingLabel={intl.formatMessage(messages.maxTeamSize)}
+                  onChange={handleChange}
+                  onBlur={handleBlur}
+                  value={values.maxTeamSize}
+                />
+                {touched.maxTeamSize && errors.maxTeamSize && (
+                  <Form.Control.Feedback type="invalid" hasIcon={false}>
+                    {errors.maxTeamSize}
+                  </Form.Control.Feedback>
+                )}
+              </Form.Group>
+            </div>
```
Fields and border spacing are not matching with figma. Helper texts are missing and do not match with figma.  Error message and helper text transition effect is not working also.

Current design:
<img width="750" alt="Screenshot 2021-09-10 at 3 02 27 PM" src="https://user-images.githubusercontent.com/79941147/132853600-37f3484d-8fb9-4380-80a8-307d08109527.png">
<img width="795" alt="Screenshot 2021-09-10 at 2 53 25 PM" src="https://user-images.githubusercontent.com/79941147/132853697-6a573a4c-f89b-420f-b02c-1a099f4552da.png">


Figma design:
<img width="428" alt="Screenshot 2021-09-10 at 5 31 54 PM" src="https://user-images.githubusercontent.com/79941147/132853653-6acfb06a-efbe-4e9d-80a3-55c012646d3d.png">
<img width="472" alt="Screenshot 2021-09-10 at 5 32 39 PM" src="https://user-images.githubusercontent.com/79941147/132853734-1ebd5620-2502-4d33-8079-68dbec124d7c.png">

awais-ansari replied:
You do not need to make changes in paragon. You can use the bootstrap class for customization which is already using in discussion topics. Please have a look at the discussion section. let me know if you need any assistance.

xitij2000 replied:
I'm not saying that making changes to paragon is a technical requirement, or that I don't know how to apply css classes. What I am considering is if the intent is to override the defaults from Pargaon (which would mean a less consistent UI across different MFEs and page) or to keep the defaults and get a more consistent appearance even if it doesn't match pixel-perfect.

xitij2000 replied:
BTW. I have updated the PR to add the transitions, and updated the text. I'm not sure what you mean by border spacing and which helper text is missing.

awaisdar001 replied:
Just wanted to share the context that pages & resources apps view did generate a bunch of tickets for bringing that closer to mock-ups. 

I think we can leave this conversation for later on during the UI/UX review.

awais-ansari commented on src/pages-and-resources/teams/Settings.jsx:
```diff
@@ -0,0 +1,150 @@
+                        onBlur={handleBlur}
+                      />
+                    ))}
+                    <Button
+                      variant="plain"
+                      onClick={() => push(blankNewTopic)}
+                    >
+                      <Icon src={Add} /> {intl.formatMessage(messages.addTeamSet)}
+                    </Button>
```
nit: spacing issue 
<img width="336" alt="Screenshot 2021-09-10 at 2 55 10 PM" src="https://user-images.githubusercontent.com/79941147/132853828-1dc88455-3f7e-4752-82ed-5fbe1f21ea15.png">

awais-ansari commented on src/pages-and-resources/teams/TeamSetEditor.jsx:
```diff
@@ -0,0 +1,208 @@
+export const TeamSetTypes = {
+  OPEN: 'open',
+  PUBLIC_MANAGED: 'public_managed',
+  PRIVATE_MANAGED: 'private_managed',
```
Move into constant file

awais-ansari replied:
There should be a constants folder that contains the whole MFE constants. It would be centralized and easy to debug.

awais-ansari commented on src/pages-and-resources/teams/TeamSetEditor.jsx:
```diff
@@ -0,0 +1,208 @@
+      <TransitionReplace>
+        {isDeleting
+          ? (
+            <div className="d-flex flex-column m-4" key="isDeleting">
+              <h3>{intl.formatMessage(messages.teamSetDeleteHeading)}</h3>
+              <p>{intl.formatMessage(messages.teamSetDeleteBody)}</p>
+              <div className="d-flex flex-row justify-content-end">
+                <Button variant="muted" size="sm" onClick={() => onDelete(teamSet)}>
+                  {intl.formatMessage(messages.delete)}
+                </Button>
+                <Button variant="muted" size="sm" onClick={cancelDeletion}>
+                  {intl.formatMessage(messages.cancel)}
+                </Button>
+              </div>
```
Delete design is not matching with Figma. Deleting function is not working properly. It has many edge cases bugs.
https://www.loom.com/share/ffd5a5ef993f4d5da93a018651ed1bc7

current design: 
<img width="775" alt="Screenshot 2021-09-10 at 5 38 51 PM" src="https://user-images.githubusercontent.com/79941147/132854437-549c409b-c075-4060-b619-881ba09dcabd.png">


Figma Design:
<img width="480" alt="Screenshot 2021-09-10 at 3 25 14 PM" src="https://user-images.githubusercontent.com/79941147/132854337-568a8344-c358-47df-b791-5794fa8bc93d.png">

awais-ansari commented on src/pages-and-resources/app-settings-modal/AppSettingsModal.jsx:
```diff
@@ -210,9 +210,9 @@ function AppSettingsModal({
+              {(formikProps.values.enabled || configureBeforeEnable) && children
               && <AppConfigFormDivider marginAdj={{ default: 0, sm: 0 }} />}
-              <AppSettingsForm formikProps={formikProps}>
+              <AppSettingsForm formikProps={formikProps} showForm={formikProps.values.enabled || configureBeforeEnable}>
```
spacing issue in footer.

current design:
<img width="841" alt="Screenshot 2021-09-10 at 2 55 20 PM" src="https://user-images.githubusercontent.com/79941147/132855249-b05437cd-1b23-4f2e-bd88-000d281a6db2.png">

Figma Design:
<img width="741" alt="Screenshot 2021-09-10 at 5 45 52 PM" src="https://user-images.githubusercontent.com/79941147/132855293-75b83261-96b2-46cc-bccf-4af848c651b6.png">

awais-ansari replied:
Reopening this. You do not need to fix this issue in paragon. These changes can be done in this MFE.

xitij2000 replied:
I'm not saying that they can't be done using CSS. Just that I was told to avoid customising the UI as much as possible so that all modals, etc look the same across the platform. If I am using a paragon component in two places, it should look the same, which won't happen if we layer styles on top of it.

awais-ansari commented on src/pages-and-resources/teams/Settings.jsx:
```diff
@@ -0,0 +1,150 @@
+    <AppSettingsModal
+      appId="teams"
+      title={intl.formatMessage(messages.heading)}
+      enableAppHelp={intl.formatMessage(messages.enableTeamsHelp)}
+      enableAppLabel={intl.formatMessage(messages.enableTeamsLabel)}
+      learnMoreText={intl.formatMessage(messages.enableTeamsLink)}
```
Team helper text is not matching with figma.
current design:
<img width="765" alt="Screenshot 2021-09-10 at 2 52 57 PM" src="https://user-images.githubusercontent.com/79941147/132855435-b0cde6f5-e15c-4fde-bf9c-38c1da1fb160.png">

Figma design:
<img width="645" alt="Screenshot 2021-09-10 at 5 47 25 PM" src="https://user-images.githubusercontent.com/79941147/132855475-4572927e-05c8-4544-b73c-b06df57e8adc.png">

awais-ansari reviewed (changes requested):
Nit: Folder naming should according to JS.

mehaknasir commented on src/pages-and-resources/teams/TeamSetEditor.jsx:
```diff
@@ -0,0 +1,174 @@
+                ) : (
+                  <div className="d-flex flex-column small flex-shrink-1">
+                    <span>{intl.formatMessage(TeamTypeNameMessage[teamSet.type].label)}</span>
+                    <span className="h4 text-truncate">{teamSet.name || '<new>'}</span>
+                    <span className="text-truncate">{teamSet.description || '<new>'}</span>
```
I think in case of new now we don't need this check as it's not collapsible so we can remove`|| '<new>'`

mehaknasir reviewed (changes requested):
Added some suggestion. kindly have a look. Thanks.

xitij2000 commented on src/pages-and-resources/teams/Settings.jsx:
```diff
@@ -0,0 +1,151 @@
+  const handleSettingsSave = async (values) => {
+    // For newly-added teams, fill in an id.
+    const topics = values.topics?.map(topic => ({
+      id: topic.id || uuid(),
```
They are called topics in the backend, so I'm using that name in the final step when we're saving to the backend.

xitij2000 commented on src/pages-and-resources/teams/Settings.jsx:
```diff
@@ -0,0 +1,151 @@
+  onClose,
+}) {
+  const [teamsConfiguration, saveSettings] = useAppSetting('teamsConfiguration');
+  const blankNewTopic = {
```
We are not adding teams here, but topics on which there can be new teams, they are called team-sets here but in the backend they are called topics.

xitij2000 commented on src/pages-and-resources/teams/Settings.jsx:
```diff
@@ -0,0 +1,151 @@
+                        teamSet={topic}
+                        errors={errors.topics?.[index]}
+                        fieldNameCommonBase={`topics.${index}`}
+                        onDelete={() => remove(index)}
```
So I think this is a cosmetic issue due to the way react renders things. I was using the index as a key, so when an item is deleted, the index is reused by another item. i.e.if you delete team2 then team3 now has the index of team2, so for a while team2 might display there till the screen is rerendered. 

It should be fixed now. If you still see this issue, do check the API call and see if it's deleting the correct item in the API.

mehaknasir replied:
this is working all fine now.

xitij2000 commented:
@mehaknasir  I've updated the PR.

mehaknasir reviewed (commented):
I have reviewed this PR. 
Two things to be noted here.

1. Is the user allowed to configure settings without enabling teams?
2. Nothing happens when I click on save. no API calls are made. can you please tell me if configuration saving is part of this PR?

xitij2000 commented:
@mehaknasir 
> Is the user allowed to configure settings without enabling teams?

Yes. In fact teams can't be enabled unless there it is configured properly. 

> Nothing happens when I click on save. no API calls are made. can you please tell me if configuration saving is part of this PR?
That is odd. This is definitely hooked up, and change any setting should be persisted using an API. 

Note that you need to be running close to master since the backend APIs for this are recent.

xitij2000 commented on src/pages-and-resources/teams/Settings.jsx:
```diff
@@ -0,0 +1,151 @@
+      description: teamSet.description,
+      max_team_size: teamSet.maxTeamSize,
+    }));
+    return saveSettings({ topics: teamSets, max_team_size: values.maxTeamSize });
```
@mehaknasir  This is here the setting are being saved. I'm checking again to see in case the recent changes broke something.

xitij2000 replied:
I've tested it, and it seems to be working.

mehaknasir replied:
Thank you for the update. I worked with another teammate he is able to save settings too. on patch call on enabling teams however we got this error. I think you might be aware of that have a look.

<img width="782" alt="Screen Shot 2021-09-15 at 2 15 14 PM" src="https://user-images.githubusercontent.com/67791278/133407543-5f8f86b5-c2be-4012-b104-af1f69bd6c1a.png">
<img width="1377" alt="Screen Shot 2021-09-15 at 2 15 33 PM" src="https://user-images.githubusercontent.com/67791278/133407556-7b667705-8d58-48af-bae9-a3fc296a5f1e.png">

xitij2000 replied:
@mehaknasir  That API needs to be updated now that the MFE is implemented. Will create a ticket for that. 
However, simply configuring teams should enable it.

mehaknasir commented on src/pages-and-resources/teams/Settings.jsx:
```diff
@@ -0,0 +1,152 @@
+                  max: TeamSizes.MAX,
+                }),
+              )
+              .default(null),
+          }),
+        )
+          .default([])
+          .uniqueProperty('name', intl.formatMessage(messages.teamSetFormNameExists)),
+      }}
```
due to missing custom error messages for these validations, we are getting such errors on the frontend. have a look. 
<img width="749" alt="Screen Shot 2021-09-15 at 2 04 54 PM" src="https://user-images.githubusercontent.com/67791278/133408071-abb9ede5-1268-448f-ad1f-23315edb2fff.png">
<img width="480" alt="Screen Shot 2021-09-15 at 2 05 42 PM" src="https://user-images.githubusercontent.com/67791278/133408080-7d442b58-fed3-4b79-9ba4-ba6fb1e30932.png">

xitij2000 replied:
I don't have a messages for these situations, but I've added a custom message for not. They can be updated with feedback from Jonathan.

xitij2000 commented:
@mehaknasir  I've updated the PR with fixes, and also renamed topics to team_sets. While the [documentation](https://edx.readthedocs.io/projects/edx-partner-course-staff/en/latest/course_features/teams/teams_setup.html#enable-and-configure-teams) still asks you to configure "topics" it seems the new name in the code is `team_sets` and topics is just around for backwards-compatibility.

mehaknasir reviewed (approved):
The frontend is reviewed and tested. Backend API for enabling/disabling pages and resources lacks support for teams. @xitij2000 will create a ticket for that. Approving frontend changes,

edx-webhook commented:
@xitij2000 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/frontend-app-course-authoring&entry.867055334=https://github.com/edx/frontend-app-course-authoring/pull/187&entry.1484655318=https://github.com/xitij2000&entry.752974735=2021-09-07+16:12&entry.1917517419=2021-09-17+09:34&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->

MHaton commented:
@xitij2000 Apologies for bugging you on here, but this is as far back as we were able to trace this new settings editor. I'm working with a fresh Palm install, and I'm unable to see this editor in Studio - is there a configuration step required for it to be displayed that we're missing? Teams appear to be enabled otherwise, as they can be configured the old bad way in Advanced Settings, but not through the Pages & Resources UI.

xitij2000 commented:
> @xitij2000 Apologies for bugging you on here, but this is as far back as we were able to trace this new settings editor. I'm working with a fresh Palm install, and I'm unable to see this editor in Studio - is there a configuration step required for it to be displayed that we're missing? Teams appear to be enabled otherwise, as they can be configured the old bad way in Advanced Settings, but not through the Pages & Resources UI.

I just had a look, and it seems that this UI is still behind a waffle flag: https://github.com/openedx/edx-platform/blob/6ab06502e0f706887f305464e3f1babe10dd73bb/lms/djangoapps/teams/waffle.py#L17

Could you try setting that and see if the UI shows up? AFAIK the waffle flag was supposed to be temporary.

MHaton commented:
> > @xitij2000 Apologies for bugging you on here, but this is as far back as we were able to trace this new settings editor. I'm working with a fresh Palm install, and I'm unable to see this editor in Studio - is there a configuration step required for it to be displayed that we're missing? Teams appear to be enabled otherwise, as they can be configured the old bad way in Advanced Settings, but not through the Pages & Resources UI.
> 
> I just had a look, and it seems that this UI is still behind a waffle flag: https://github.com/openedx/edx-platform/blob/6ab06502e0f706887f305464e3f1babe10dd73bb/lms/djangoapps/teams/waffle.py#L17
> 
> Could you try setting that and see if the UI shows up? AFAIK the waffle flag was supposed to be temporary.

That did it - you're absolutely right, it was the waffle flag I was missing! Thanks for letting me know, much appreciated.
